File: Python/0061-m-rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

"""
Linked List
"""
class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if not head:
            return None
        
        curt = head
        size = 0
        while curt:
            size += 1
            curt = curt.next
        
        k = k % size
        if k == 0:
            return head
        
        slow = fast = head
        for _ in range(k):
            fast = fast.next
        while fast.next:
            slow = slow.next
            fast = fast.next
        
        newHead = slow.next
        fast.next = head
        slow.next = None
        return newHead

File: leetcode/C++/0061-m-rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (head == nullptr) return nullptr;

        ListNode* cur = head;
        int size = 0;
        while (cur != nullptr) {
            size++;
            cur = cur->next;
        }

        k = k % size;
        if (k == 0) return head;

        ListNode* slow = head;
        ListNode* fast = head;
        for (int i = 0; i < k; i++)
            fast = fast->next;
        while (fast->next != nullptr) {
            slow = slow->next;
            fast = fast->next;
        }

        ListNode* newHead = slow->next;
        fast->next = head;
        slow->next = nullptr;
        return newHead;
    }
};